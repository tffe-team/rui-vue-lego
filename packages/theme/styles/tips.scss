.r-vue-lego-tips{
  .icon-box{
    .icon{
      font-size: .14rem;
    }
    margin-bottom: .1rem;
  }
}

