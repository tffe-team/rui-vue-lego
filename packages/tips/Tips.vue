<template>
  <div :class="['r-vue-lego-tips', iconName ? 'vue-lego-tips-icon' : '']" v-show="visible">
    <div class="icon-box" v-if="iconName"> 
      <i class="rui-ico icon" v-if="iconName==='success'">&#xe60b;</i>
      <i class="rui-ico icon" v-if="iconName==='fail'">&#xe90b;</i>
    </div>
    <div class="message">{{message}}</div>
  </div>
</template>
<script>
export default {
  name: 'r-tips',
  props: {
    iconName : {
      type : String,
      required: false
    },
    message : {
      type : String,
      required: true
    }
  },
  data() {
    return {
      visible : this.value
    }
  }
}
</script>
<style lang="scss" scoped>
.r-vue-lego-tips{
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  min-width: 2rem;
  max-width: 72%;
  z-index: 9999;
  padding: 0.1rem;
  border-radius: 0.04rem;
  background-color: rgba(0, 0, 0, 0.6);
  color: #fff;
  text-align: center;
  min-width: 1.5rem;
  .icon{
    font-size: .2rem;
  }
}
.vue-lego-tips-icon{
  min-width: 1rem;
}
</style>