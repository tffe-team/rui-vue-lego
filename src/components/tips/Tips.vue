<template>
  <div :class="['r-vue-lego-tips', 'vue-lego-tips', iconName ? 'vue-lego-tips-icon' : '', 'vue-lego-fade-out']" v-show="visible">
    <div class="icon-box" v-if="iconName"> 
      <i class="rui-ico icon" v-if="iconName==='success'">&#xe60b;</i>
      <i class="rui-ico icon" v-if="iconName==='fail'">&#xe90b;</i>
    </div>
    <div class="message">{{message}}</div>
  </div>
</template>
<script>
export default {
  name: 'r-tips',
  props: {
    iconName : {
      type : String,
      required: false
    },
    message : {
      type : String,
      required: true
    }
  },
  data() {
    return {
      visible : this.value
    }
  }
}
</script>