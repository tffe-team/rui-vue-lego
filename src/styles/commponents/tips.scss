@include rui-tips(
){
  min-width: 1rem;
};
.r-vue-lego-tips {
  z-index: 1000;
  .icon-box {
    .icon {
      font-size: .2rem;
    }
  }
}