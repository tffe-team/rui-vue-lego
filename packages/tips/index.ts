import Vue from 'vue'
import Tips from './Tips.vue'
import { ToastOption } from '../../typings/lego'

let tipstInstance

const $tips = properties => {
  const _props = Object.assign({
    onRemove: function () {   
    },
    onClose: function() {
    }
  }, properties)
  let timer = null
  const onClose =  _props.onClose
  const onRemove =  _props.onRemove
  delete _props.onClose
  delete _props.onRemove
  if(!properties.onRemove) {
    createToast(properties)
    return 
  }
  const Instance = new Vue({
    data: Object.assign(_props, {
      duration: 3000
    }),
    mounted() {
      this.close()
    },
    methods: {
      close() {
        timer = setTimeout(() => {
          this.$children[0].visible = false
          this.remove()
          onClose()
        }, this.duration)
      },
      remove() {
        if (this.$el) {
          document.body.removeChild(this.$el)
        }
        clearTimeout(timer)
        onRemove()
      }
    },
    render (h: any) {
      return h(Tips, {
        props: Object.assign({}, _props),
        domProps: {
          value: this.visible
        },
        on: {
          'input': (status) => {
            this.visible = status
          }
        }
      })
    }
  }).$mount()
  document.body.appendChild(Instance.$el)
  const tips = Instance.$children[0]
  return {
    show() {
      if(tips.visible) {
        this.remove()
      } else {
        tips.visible = true
      }
      tips.visible = true
    },
    remove() {
      tips.visible = false
      tips.$parent.remove()
    },
    component: tips
  }
}

function getToastInstance(options) {
  tipstInstance = tipstInstance ||  $tips(options)
  return tipstInstance
}
function createToast(options) {
  options.onRemove = function() {
    tipstInstance = null
  }
  getToastInstance(options)
  tipstInstance.show()
}
$tips.info = (option: ToastOption) => {
  createToast(option)
}
$tips.success = (option: ToastOption) => {
  option.iconName = 'success'
  createToast(option)
}
$tips.error = (option: ToastOption) => {
  option.iconName = 'fail'
  createToast(option)
}
export default $tips