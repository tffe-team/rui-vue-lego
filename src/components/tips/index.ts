import Vue from 'vue'
import Tips from './Tips.vue'
import { TipsOption } from '../../../typings/vue-lego'

let tipstInstance

const $tips = properties => {
  const _props = Object.assign({
    onRemove: function () {   
    },
    onClose: function() {
    }
  }, properties)
  let timer = null
  const onClose =  _props.onClose
  const onRemove =  _props.onRemove
  delete _props.onClose
  delete _props.onRemove
  if(!properties.onRemove) {
    createTips(properties)
    return 
  }
  const Instance = new Vue({
    data: Object.assign({
      duration: 3000
    }, _props),
    mounted() {
     this.close()
    },
    methods: {
      close() {
        timer = setTimeout(() => {
          this.$children[0].visible = false
          this.remove()
          onClose()
        }, this.duration)
      },
      remove() {
        if (this.$el) {
          document.body.removeChild(this.$el)
        }
        clearTimeout(timer)
        onRemove()
      }
    },
    render (h: any) {
      return h(Tips, {
        props: Object.assign({}, _props),
        domProps: {
          value: this.visible
        },
        on: {
          'input': (status) => {
            this.visible = status
          }
        }
      })
    }
  }).$mount()
  document.body.appendChild(Instance.$el)
  const tips = Instance.$children[0]
  return {
    show() {
      if(tips.visible) {
        this.remove()
      } else {
        tips.visible = true
      }
      tips.visible = true
    },
    remove() {
      tips.visible = false
      tips.$parent.remove()
    },
    component: tips
  }
}

function getTipsInstance(options) {
  tipstInstance = tipstInstance ||  $tips(options)
  return tipstInstance
}
function createTips(options) {
  options.onRemove = function() {
    tipstInstance = null
  }
  getTipsInstance(options)
  tipstInstance.show()
}
$tips.info = (option: TipsOption) => {
  createTips(option)
}
$tips.success = (option: TipsOption) => {
  option.iconName = 'success'
  createTips(option)
}
$tips.error = (option: TipsOption) => {
  option.iconName = 'fail'
  createTips(option)
}
export default $tips