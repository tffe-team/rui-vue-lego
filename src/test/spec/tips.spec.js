import $tips from '@components/tips'
import {later} from '../index'

const selector = '.r-vue-lego-tips'

test('create a tips', async () => {
  $tips.info({
    message: 'create a tips',
  })

  await later()
  expect(document.body.querySelector(selector)).toMatchSnapshot()
  await later(3000)
})

test('tips type', async() => {
  $tips.success({
    message: 'a success tips',
  })

  await later()
  expect(document.body.querySelector(selector)).toMatchSnapshot()
  await later(3000)
})

test('auto close', async () => {
  $tips.info({
    message: 'auto close',
  })

  await later()
  expect(document.body.querySelector(selector)).toBeTruthy()
  await later(3000)
  expect(document.body.querySelector(selector)).toBeFalsy()
})
