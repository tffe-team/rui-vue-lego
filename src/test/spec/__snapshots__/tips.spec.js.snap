// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a tips 1`] = `
<div
  class="r-vue-lego-tips vue-lego-tips vue-lego-fade-out"
  style=""
>
  <!---->
   
  <div
    class="message"
  >
    create a tips
  </div>
</div>
`;

exports[`tips type 1`] = `
<div
  class="r-vue-lego-tips vue-lego-tips vue-lego-tips-icon vue-lego-fade-out"
  style=""
>
  <div
    class="icon-box"
  >
    <i
      class="rui-ico icon"
    >
      î˜‹
    </i>
     
    <!---->
  </div>
   
  <div
    class="message"
  >
    a success tips
  </div>
</div>
`;
